Anarchy

asymmetric opened the pull request:
This PR strips out all governance actions.

If governance is malicious, it will always `burn` and never `free`.

If it isn't, it can `mint` after the fact.

This is a tradeoff though: `mint`ing requires some off-chain coördination (to get the list of ESM pledgers) and is more complex than `exit`'s "pull" model.

Closes #1, closes #6.

asymmetric committed:
remove governance actions

Governance actions (`free`, `burn`) can be completely eliminated if we
make the following assumptions:

* if governance is malicious, they'll never `free` after `fire`, and
always attempt to `burn` before `fire`
* if governance is not malicious, they can `mint` new tokens after
`fire`

asymmetric committed:
remove states in favor of fired boolean

asymmetric committed:
cap -> min

It's not really a cap as there's no upper limit.

asymmetric committed:
update README

andy8052 commented on src/ESM.sol:
```diff
@@ -66,67 +52,26 @@ contract ESM is DSNote {
-    }
-
-    // -- user actions --
     function join(uint256 wad) external note {
```
What do you think about having `require(!full())` here now. This could help to reduce the off chain complexity in the case where gov wants to mint new tokens for people.

asymmetric replied:
Honestly I don't know if it would really make a difference. If there's lots of small donations then it will be a PITA anyway, even just for the transactions below the threshold; if there's only a few large ones then they're likely coordinated and will likely use a multisig and do a `join+fire` from there.

andy8052 replied:
Yeah good point

andy8052 commented on src/ESM.sol:
```diff
@@ -66,67 +52,26 @@ contract ESM is DSNote {
     function fire() external note {
-        require(state == START && full(), "esm/not-fireable");
+        require(!fired, "esm/already-fired");
+        require(sum >= min, "esm/min-not-reached");
```
Why did you move away from calling `full` here?

asymmetric replied:
Because I wanted to make `full() external`, for the reasons Lev brought up some time ago, namely that it makes the bytecode easier to audit/FV than `public`.

I know it looks weird, so I'm open to removing `fuil`.

MrChico replied:
There's not much difference in the difficulty of FVing `public` vs. `external` functions

asymmetric replied:
OK I have no clue about this, but @livnev seemed to be into the idea.

I'd love to stop cargo-culting! :D

livnev replied:
@asymmetric @MrChico My thinking is along the lines of: (i) change `public` to `external` whenever possible, at the very least to reduce bytecode complexity; (ii) in some cases, if you really need to call it internally, consider changing `public` to `internal`, since then it will be easier to FV the calling methods using the internal method's spec as a lemma; (iii) in some cases, consider removing the method altogether.

Applying the analysis to this situation makes me want to ask: do we need `full()` at all?

MrChico replied:
@livnev `external` vs. `public` only makes a difference for functions whose arguments cannot be stored directly on the stack, i.e. for dynamic types. In those cases, an `external` function can keep the arguments in `calldata` while a `public` function will have to copy them to `memory`. For value types, the difference between an `public` and `external` function are fairly equivalent.

livnev replied:
@MrChico I agree. There is also a general theme of trying to limit the number of entry points to your contract and making it so that a piece of code is not executed as part of too many paths (since this increases the overhead of reasoning about correctness). In this simple case I agree that it doesn't make a difference one way or the other.

asymmetric replied:
OK, seems like switching back to `public` is OK in this case.

asymmetric committed:
make full public

It doesn't make FV harder, since the function has no (reference-type) arguments.

gbalabasquer commented on src/ESM.sol:
```diff
@@ -66,67 +52,26 @@ contract ESM is DSNote {
-        bool ok = gem.transfer(usr, wad);
-
-        require(ok, "esm/failed-transfer");
+        require(gem.transferFrom(msg.sender, address(0x0), wad), "esm/transfer-failed");
```
I know for the `flap` auctions we are still sending to `address(0)`, but that will eventually change, it is not the pattern we are following for burning the MKR token, as `address(0)` will not affect the `totalSupply`.
I think we have 2 options: leaving the burner address as param or use directly `burn` function (which has to be previously authorized and I don't like much).

asymmetric replied:
Yeah agreed, I don't like the idea of `rely`ing the ESM on the MKR token.

Don't know what the current burner contract looks like (IIRC there's one), but I think it should have a publicly accessible `burn` function, so that `burn`ing is not up to governance.

gbalabasquer commented on src/ESM.sol:
```diff
@@ -66,67 +52,26 @@ contract ESM is DSNote {
-        require(state == START, "esm/not-joinable");
+        require(!fired, "esm/already-fired");
 
         gems[msg.sender] = add(gems[msg.sender], wad);
```
Btw do we really need `gems` at all? if it will be done offchain we could still get those addresses which sent funds.

asymmetric replied:
Seems like having `gems` would make it easier to audit who pledged funds, without having to run a full node, and therefore to ensure that the token redistribution is fair.

gbalabasquer replied:
Yeah true, I was thinking we could use the events or the different txs to get these values, but yeah, having the gems is probably more practical.

gbalabasquer reviewed (commented):
Are you adding the `ESMFab`?

asymmetric commented:
@gbalabasquer as discussed, we don't need the ESFab, everything will be handled in a spell.

desaperados commented:
For consistency with naming in other `dss` modules consider changing `gems[user]` to non-plural `gem[user]`.

e.g.

* `pie[user]` - Dai Savings user balance
* `Pie` - Total Dai Savings
---
* `sin[user]` - Bad debt user balance
* `vice` - Total bad debt 
---
* `dai[user]` - Dai user balance
* `debt` - Total Dai 

Or event better, avoid any possible confusion with `Vat.gem[ilk, user]` consider re-naming `gems` completely.

* `sum[user]` - MKR user balance
* `Sum` - Total MKR

asymmetric committed:
reintroduce sun

Transferring to 0x0 prevents us from keeping the total supply up to
date. Instead, we will use a publicly callable burner contract.

asymmetric committed:
gems -> Sum

To mirror the usage in dss (e.g. line/Line).

asymmetric committed:
EndTest -> EndMock

More descriptive

asymmetric commented:
I find `gem` in the Vat to be a strange choice. `Sum` should also make the distinction with `sun` easier, as those are the only two scalar types, I've gone with that. Not super sure it's less confusing though.

gbalabasquer commented:
@asymmetric I guess we are ready to merge this, right?
